apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -21,28 +30,142 @@ def get_file_names_as_array(raw_file_names: str) -> list[str]:
+    returns the set of file names that match any of the given patterns.
+    """
+    result = set()
+    for pattern in regex_patterns:
+        regex = re.compile(pattern[0], re.IGNORECASE)
+        for file in files_names:
+            if regex.match(file):
+                result.add(file)
+
+    return result
+
+
+def extract_patterns(optional_files: list[str]) -> tuple[list[list[str]], list[str]]:
+    """
+    Takes a list of file names and returns it split into wildcard patterns and plain names
+
+    File names with with wildcards are converted into regular expressions and returned as a
+    list of two-item lists, each containing both the regex for matching and the plain pattern
+    for display. Plain file names are returned in a list with escapes removed for simple
+    string-based comparison
+    """
+    optional_files_plain = []
+    optional_files_regex = []
+
+    for file in optional_files:
+        # glob_to_regex returns an empty string if the file name contains no wildcard
+        regex_raw = glob_to_regex(file)
+        if regex_raw:
+            optional_files_regex.append([regex_raw, file])
+        else:
+            optional_files_plain.append(file.replace("\\", ""))
+
+    return optional_files_regex, optional_files_plain
+
+
+def glob_to_regex(glob_pattern: str) -> str:
+    """
+    Translates a glob pattern into a regex that can be handled consistently across Python and JS
+    Returns the regex if wildcards are present, and an empty string otherwise
+    """
+    result = ["^"]
+    has_wildcard = False
+    escape = False
+    in_range = False
+    for c in glob_pattern:
+        if escape:
+            result.append(re.escape(c))
+            escape = False
+        elif in_range and c != "]":
+            result.append(c)
+        elif c == "\\":
+            escape = True
+        elif c == "?":
+            has_wildcard = True
+            result.append(".")
+        elif c == "*":
+            has_wildcard = True
+            result.append(".*")
+        elif c == "[":
+            has_wildcard = True
+            in_range = True
+            result.append(c)
+        elif c == "]":
+            in_range = False
+            result.append(c)
+        else:
+            result.append(re.escape(c))
+
+    # If there are no wildcards, return an empty string
+    if not has_wildcard:
+        return ""
+
+    result.append("$")
+    return "".join(result)
+
+
+def get_answer_name(file_names: str, optional_file_names: str = "") -> str:
+    """
+    Computes the unique identifer of a pl-file-upload element, which is the SHA1 hash of its
+    file-names and optional-file-names attributes
+    """
+    # Using / as separator as the only character guaranteed not to appear in file names
+    combined_name = file_names + (
+        "/" + optional_file_names if optional_file_names else ""
+    )
+
     return "_file_upload_{0}".format(
```

Comment:
```suggestion
    return "_file_upload_" + hashlib.sha1(combined_name.encode("utf-8")).hexdigest()
```
Nit: Slightly shorter. The use of `.format` is also a little outdated, should remove usages from this file and use either the above or f-strings instead.